src/v/pandaproxy/kafka_client_cache.cc

```diff
@@ -0,0 +1,175 @@
+ss::future<> stop_clients_until_empty(List_Type& list, mutex& lock) {
+    while (!list.empty()) {
+        co_await lock.with([&list] { return list.back().client->stop(); });
+        list.pop_back();
+    }
+}
+} // namespace
+
+kafka_client_cache::kafka_client_cache(
+  YAML::Node const& cfg,
+  config::binding<size_t> max_size,
+  config::binding<std::chrono::milliseconds> keep_alive)
+  : _config{cfg}
+  , _cache_max_size{max_size}
+  , _keep_alive{keep_alive} {}
+
+client_ptr kafka_client_cache::make_client(
+  credential_t user, config::rest_authn_method authn_method) {
+    kafka::client::configuration cfg{
+      to_yaml(_config, config::redact_secrets::no)};
+
+    // Set the principal when the request is using HTTP Basic AuthN
+    if (authn_method == config::rest_authn_method::http_basic) {
+        // Need to specify type or else bad any_cast runtime error
+        cfg.sasl_mechanism.set_value(ss::sstring{"SCRAM-SHA-256"});
+        cfg.scram_username.set_value(std::move(user.name));
+        cfg.scram_password.set_value(std::move(user.pass));
+    }
+
+    return ss::make_lw_shared<kafka::client::client>(
+      to_yaml(cfg, config::redact_secrets::no));
+}
+
+ss::future<client_ptr> kafka_client_cache::fetch_or_insert(
+  credential_t user, config::rest_authn_method authn_method) {
+    auto& inner_list = _cache.get<underlying_list>();
+    auto& inner_hash = _cache.get<underlying_hash>();
+    ss::sstring k{user.name};
+    auto it_hash = inner_hash.find(k);
+
+    client_ptr client;
+
+    // When no client is found ...
+    if (it_hash == inner_hash.end()) {
+        if (_cache.size() >= _cache_max_size()) {
+            if (!_cache.empty()) {
+                auto item = inner_list.back();
+                vlog(plog.debug, "Cache size reached, evicting {}", item.key);
+                inner_list.pop_back();
+                _evicted_items.push_back(item);
+            }
+        }
+
+        vlog(plog.debug, "Make client for user {}", k);
+
+        client = make_client(user, authn_method);
+        inner_list.emplace_front(k, client);
+    } else {
+        co_await _lock.with([this,
+                             k,
+                             user,
+                             authn_method,
+                             &it_hash,
+                             &inner_hash] {
+            // Make a new client if the passwords don't match
+            if (it_hash->client->config().scram_password.value() != user.pass) {
+                vlog(
+                  plog.debug, "Password changed. Remake client for user {}", k);
+                client_ptr old = it_hash->client;
+                client_ptr next = make_client(user, authn_method);
+                // Try connect as a safety check
+                return next->connect().then_wrapped(
+                  [this, k, next, old, &it_hash, &inner_hash](ss::future<> f) {
+                      if (!f.failed()) {
+                          inner_hash.modify(
```

Comment:
Spoke with @dotnwat about this. Related to https://github.com/redpanda-data/redpanda/pull/6452#discussion_r982919802
Since we don't know about BMI implementation, we can't be certain .modify() is safe. Need to research more.

Reply:
i think it really all boils down to how much concurrency control you want to add.